src/main/java/com/mongodb/hibernate/service/MongoClientCustomizer.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.service;
+
+import com.mongodb.MongoClientSettings;
+import org.hibernate.service.Service;
+
+/**
+ * A Hibernate {@link Service} focusing on customizing {@link com.mongodb.client.MongoClient} creation programmatically
+ * by invoking the various methods of {@link MongoClientSettings.Builder} before its
+ * {@link MongoClientSettings.Builder#build()} invocation.
+ *
+ * <p>Note that this service is only needed when normal connection string, user and password configurations are not
+ * enough.
+ *
+ * <p>An example usage is as follows:
+ *
+ * <pre>
+ * var cfg = new Configuration();
+ *
+ * // configure cfg as you normally do (e.g. add entity classes)
+ * ...
+ *
+ * // connection string, user and password have been applied to the builder already
+ * var clientCustomizer = builder -> {
+ *     // customize client settings to your heart's content
+ *     ...
+ * };
+ *
+ * var serviceRegistryBuilder = cfg.getStandardServiceRegistryBuilder();
+ * serviceRegistryBuilder.addService(MongoClientCustomizer.class, clientCustomizer);
+ *
+ * var sessionFactory = cfg.buildSessionFactory();
+ *
+ * // start using sessionFactory as normal
+ * ...
+ *
+ * </pre>
+ *
+ * <p>The injected {@code clientCustomizer} service will be applied at later stage (in
+ * {@link com.mongodb.hibernate.jdbc.MongoConnectionProvider}) of Hibernate bootstrapping.
+ *
+ * @see com.mongodb.hibernate.jdbc.MongoConnectionProvider#configure(java.util.Map)
+ */
+@FunctionalInterface
+public interface MongoClientCustomizer extends Service {
+    void customize(MongoClientSettings.Builder builder);
```

Comment:
`ConnectionString` and `MongoClientSettings` conflict with each other, because they are different ways of configuring mostly the same settings. The Java driver avoids coming up with rules on how conflicts are resolved by shifting the task of conflict resolution to an application: an application has to call `MongoClientSettings.Builder.applyConnectionString`, and has a choice to configure `MongoClientSettings.Builder` before this call or after (or even both).

Let's use the same approach here by changing the method to
```java
void customize(MongoClientSettings.Builder mongoClientSettingsBuilder, @Nullable ConnectionString connectionString)
```

This way `MongoClientCustomizer` will be able to decide when `applyConnectionString` should be called, if at all. Also, you may see that the `connectionString` argument is `@Nullable`. I think, we should make `JdbcSettings.JAKARTA_JDBC_USER` optional, because if someone wants to configure `MongoClient` fully programmatically, we must not interfere with that by demanding a connection string to be specified.

Reply:
Too many options. We'd better finalize the tech design prior to implementation?

For the second option accepting a map, why bother with passing them from `hibernate.property` intermediary (I assume most of the map entries are irrelevant), instead of letting user build a client from scratch in whatever way he/she wants? This is a programattic way user has maximal flexibility, right?

Reply:
> if we do allow user/password (or anything else for that matter) to be specified via configuration, how would those values get to the customizer?

A setting that can be specified via a connection string should not be allowed to be specified via Hibernate ORM configuration properties. That way we won't have the problem you pointed out. I fail to see a single benefit of allowing connection string options to be also specified via Hibernate ORM configuration properties. If someone sees those benefits, could they be stated explicitly?

> We'd better finalize the tech design prior to implementation?

That's why when I created [HIBERNATE-38 Design, implement and document how an application can configure the product](https://jira.mongodb.org/browse/HIBERNATE-38), I specified it as a dependency of https://jira.mongodb.org/browse/HIBERNATE-28 (the task implemented by the current PR).

Reply:
Let me provide some historical context. The decision to make use of the following 3 properties as normal way to bootstrap our product was made by @jyemin:

- jakarta.persistence.jdbc.url
- jakarta.persistence.jdbc.user
- jakarta.persistence.jdbc.password

I want to point out that though the above JPA properties are preferred, Hibernate also has its own legacy property counterparts, including:

- hibernate. connection. url
- hibernate. connection. username
- hibernate. connection. password

together with the other multi-db related default catalog and schema.
So a natural question is if our user used the properties from `Environment`, they migtht be surprised that they are not taking effect at all. If we chose `default catalog` and `default schema`, they are even more surprised that they are supported, :).

My personal preference is to avoid them (they are for traditional scenario that a JDBC Driver is provided together with them (another `jakarta.persistence.jdbc.driver` is usually provided together with above three; but our case is definitively special in the sense that we have no JDBC Driver to provide) and choose our own properties (like Spring Data MongoDB's config) so end-user would have crytal clear expectation how to use our product.

From my understanding, if we wanna avoid the multi-db complexity (it is complicated for in native MQL scenario, a default database is still needed, and tbh, I don't think reusing the confusing `catalog` and `schema` is a great product idea in the first place), our product is only expecting two things and nothing more:

- give me a `MongoClient`
- give me a mongo database name

If so, a simplistic tech design is simply:

- introduce a new property (mandatory) like `mongodb.client`, which will accept a working `MongoClient` prepared by end user
- introduce a new property (mandatory) like `mongodb.database`, which denotes the database name

I think even `Service` is not necessary and adding the `mongoClient` into a settings map is perfect fine for now (just the original idea @jyemin previously proposed; in the future when we need to consider using our product in JavaEE environment where it is not easy to insert ad-hoc code to create a `MongoClient`, we could switch to `Service` mechanism).

Reply:
> I like this idea, but if we do allow user/password (or anything else for that matter) to be specified via configuration, how would those values get to the customizer?

That is straightforward for we have gotten a mandatory `ConnectionString` already (in `MongoConnectionProvider`). We only need to tweak the logic to make it optional in event of customizer is found.

Reply:
> That's why when I created [HIBERNATE-38 Design, implement and document how an application can configure the product](https://jira.mongodb.org/browse/HIBERNATE-38), I specified it as a dependency of https://jira.mongodb.org/browse/HIBERNATE-28 (the task implemented by the current PR).

I am ok with the discussion. My only concern is it might be a lenghy process and might interfere with Milestone 1's progress (if the resulting conclusion is easy to implement, that is also fine). We are still at the very first ticket, but we have found many reasons to digress. Sometimes sticking to a simple solution pays off (no matter whther it is perfect, as long as it could be rectified later), and that is why we spent so much time on tech doc writing. 

Another option is to move the discuss to Milestone 2 and stick to the current approach based on the three JPA Hibernate properties? I am even more versed in the details and complexity, but as I explained, we'd better focus.